### PR TITLE
fix to upload folder path

### LINKED ISSUE #194: Update upload lambda to use filename to create upload path
Link to code that needs update: https://github.com/IMAP-Science-Operations-Center/sds-data-manager/blob/1f6109cd03eb35d773e0b6336627cff0785efe79/sds_data_manager/lambda_code/SDSCode/upload_api.py#L56

### PR DESCRIPTION
**Pull request by tech3371**
# Change Summary

## Overview
Fix upload folder path

## Updated Files
- sds_data_manager/config/config.json
   - update and added new product
- sds_data_manager/lambda_code/SDSCode/upload_api.py
   - Added fixes for upload folder path

### CODE REVIEW

File: sds_data_manager/lambda_code/SDSCode/upload_api.py
Comment (greglucas) [inline sds_data_manager/lambda_code/SDSCode/upload_api.py]: Do we need this new function or could you use this to replace the `check_for_matching_filetype` function above this? Since we have such a rigid structure now, I'm not sure we even need the config dictionary file.

```python
mission, instrument, data_level, descriptor, startdate, enddate = filename.split("_")
# create your path from these, maybe raise if something doesn't match our requirements
```

Comment (tech3371) [reply sds_data_manager/lambda_code/SDSCode/upload_api.py]: `check_for_matching_filetype` checks that the filename follows one of the pattern in the `config.json`. Once it finds that format, then it stops there and continue. If we remove `check_for_matching_filetype`, then having `config.json` is not useful.

Comment (greglucas) [reply sds_data_manager/lambda_code/SDSCode/upload_api.py]: Yeah, so up to you whether you think we'll need that check/config, but now that we are more rigid in the filename requirements I'm not so sure we will was my suggestion to merge them together. Adding onto my above suggestion we could add checks from the config file directly here as well:
```python
mission, instrument, data_level, descriptor, startdate, enddate = filename.split("_")
# create your path from these, maybe raise if something doesn't match our requirements
if not enddate.endswith((".pkts", ".cdf")):
    raise ValueError("Bad filetype, the SDC requires '.pkts' or '.cdf' filetypes to be uploaded")
if instrument not in (...):
    raise ValueError("Acceptable instruments are: ...")
# Even more validation on things if you want as well.
file_path_to_upload = same_as_you_have_below
```

File: sds_data_manager/lambda_code/SDSCode/upload_api.py
Comment (bourque) [inline sds_data_manager/lambda_code/SDSCode/upload_api.py]:
```suggestion
    path_to_upload_file : str
```

File: sds_data_manager/lambda_code/SDSCode/upload_api.py
Comment (bourque) [inline sds_data_manager/lambda_code/SDSCode/upload_api.py]: I suggest to add this info to the function docstring rather than an inline comment here. I think then it could be more easily referenced within the sphinx docs.

**Review by bourque — DISMISSED** [review]
Looks good! I just had a few non-blocking suggestions.

**Comment by tech3371** [comment]
@greglucas and @maxinelasp Can you review the `path_helper.py` to see if I captured your suggestions? After your approval, I will work towards adding tests and incorporate into `indexer.py` too.

File: sds_data_manager/config/config.json
Comment (greglucas) [inline sds_data_manager/config/config.json:1]: I think we were going to remove this file. Another benefit I thought of is that we won't need to download this onto the Lambda every single time if we put the configuration directly into the code.

File: sds_data_manager/lambda_code/SDSCode/path_helper.py
Comment (greglucas) [inline sds_data_manager/lambda_code/SDSCode/path_helper.py:9]: I sort of worry we are going too class-heavy on IMAP... Is there a reason we need this stored/instantiated as an object versus using chains of functions? I think you could chain a bunch of functions together and it would be easier to re-use for other purposes too then.

My personal preference is to use global constants for your valid options and then check against those:
```python
VALID_INSTRUMENTS = {"codice", "glows", ...}
# Could make this more robust later and restrict it to specific data levels for each instrument
VALID_DATALEVELS = {"l0", ...}

def is_filename_valid(filename):
    instrument, datalevel, ... = filename.split("_")
    if instrument not in VALID_INSTRUMENTS:
        raise ValueError(f"{instrument} is not a valid instrument on IMAP, must be one of {VALID_INSTRUMENTS})
```

You could also break those checks out into individual functions if you wanted too.

File: sds_data_manager/lambda_code/SDSCode/path_helper.py
Comment (greglucas) [inline sds_data_manager/lambda_code/SDSCode/path_helper.py]: I don't trust that people won't put `.` in the middle of their data product names... I guess we could enforce that, but I think we could handle the underscores alone, and then strip off the extension later.
```suggestion
        self.split_filename = filename.split("_")
```

File: sds_data_manager/lambda_code/SDSCode/path_helper.py
Comment (greglucas) [inline sds_data_manager/lambda_code/SDSCode/path_helper.py]: Do you need this, or can you just let your try/except below catch this case too?

File: sds_data_manager/lambda_code/SDSCode/path_helper.py
Comment (greglucas) [inline sds_data_manager/lambda_code/SDSCode/path_helper.py:195]: Rather than just creating the path to the upload, should we just create the entire s3 uri here? We have the filename too, so I think that would save us the headache later of having to append the filename to this path.

File: sds_data_manager/lambda_code/SDSCode/path_helper.py
Comment (maxinelasp) [reply sds_data_manager/lambda_code/SDSCode/path_helper.py]: Agree with Greg, below will catch only 8 digit times

File: sds_data_manager/lambda_code/SDSCode/upload_api.py
Comment (maxinelasp) [inline sds_data_manager/lambda_code/SDSCode/upload_api.py]: Is this intentionally included? Or just for testing?

File: sds_data_manager/lambda_code/SDSCode/path_helper.py
Comment (maxinelasp) [reply sds_data_manager/lambda_code/SDSCode/path_helper.py:9]: I like the class, personally. That being said, I would probably prefer to pass it in to the functions that use it as a parameter so other users could override it easily. (eg in line 114, maybe you could have `file_pattern_config` as an optional keyword and default to this FilePatternConfig class.)

**Review by maxinelasp — DISMISSED** [review]
I like your changes, they look good to me. If config.json isn't needed anymore, then it should be removed before merging.

File: sds_data_manager/config/config.json
Comment (tech3371) [reply sds_data_manager/config/config.json:1]: Yes, True.
Yeah, we will need to remove this file. I didn't do that in this PR yet, because removing this is going to effect other CDK stack and I didn't want to bring in more changes in this PR. I plan to remove this and stacks that are associated with this in upcoming PR:
* S3 bucket stack
* resources that uploads this file
* and may be others as well that I don't about.

File: sds_data_manager/lambda_code/SDSCode/path_helper.py
Comment (tech3371) [reply sds_data_manager/lambda_code/SDSCode/path_helper.py:9]: I do like Class for this case because it groups all associated parameters together and easy to access.

So @maxinelasp Are you suggesting something like this?
```
    def validate_filename(self, file_pattern_config=FilenamePatternConfig()) -> bool:

```

File: sds_data_manager/lambda_code/SDSCode/path_helper.py
Comment (tech3371) [reply sds_data_manager/lambda_code/SDSCode/path_helper.py]: ok. will do. Changing to this:
```
        self.split_filename = filename.split("_")
        (
            self.mission,
            self.instrument,
            self.data_level,
            self.descriptor,
            self.startdate,
            self.enddate,
            self.last_value
        ) = self.split_filename
        (self.version, self.extension) = self.last_value.split(".")
```

File: sds_data_manager/lambda_code/SDSCode/path_helper.py
Comment (tech3371) [reply sds_data_manager/lambda_code/SDSCode/path_helper.py]: cool. I will remove this.

File: sds_data_manager/lambda_code/SDSCode/path_helper.py
Comment (tech3371) [reply sds_data_manager/lambda_code/SDSCode/path_helper.py:195]: Will do that.

File: sds_data_manager/lambda_code/SDSCode/upload_api.py
Comment (tech3371) [reply sds_data_manager/lambda_code/SDSCode/upload_api.py]: yes. just for testing. good catch!

File: sds_data_manager/lambda_code/SDSCode/path_helper.py
Comment (tech3371) [reply sds_data_manager/lambda_code/SDSCode/path_helper.py:9]: I push new code with these changes. I hope I address your comments.

Comment (maxinelasp) [reply sds_data_manager/lambda_code/SDSCode/path_helper.py:9]: Yes, that's what I had in mind!

**Comment by tech3371** [comment]
@bourque @maxinelasp @greglucas Can you review it again? No rush. I have to ping because this repo undoes your approval.

File: pyproject.toml
Comment (greglucas) [inline pyproject.toml]: I don't think we should avoid this error.

File: sds_data_manager/lambda_code/SDSCode/path_helper.py
Comment (greglucas) [inline sds_data_manager/lambda_code/SDSCode/path_helper.py]: The below strptime should take care of all the regex work for you.
```suggestion
```

File: sds_data_manager/lambda_code/SDSCode/path_helper.py
Comment (greglucas) [inline sds_data_manager/lambda_code/SDSCode/path_helper.py]:
```suggestion
    def validate_filename(self, file_pattern_config=None) -> bool:
        ...
        if file_pattern_config is None:
            file_pattern_config = FilenamePatternConfig()
```

File: sds_data_manager/lambda_code/SDSCode/path_helper.py
Comment (greglucas) [inline sds_data_manager/lambda_code/SDSCode/path_helper.py]:
```suggestion
```

File: sds_data_manager/lambda_code/SDSCode/path_helper.py
Comment (maxinelasp) [inline sds_data_manager/lambda_code/SDSCode/path_helper.py]: This can just be removed, as the `strptime` will also fail if the input date isn't 8 digits exactly.

**Review by maxinelasp — APPROVED** [review]

File: tests/lambda_endpoints/test_path_helper.py
Comment (greglucas) [inline tests/lambda_endpoints/test_path_helper.py]:
```suggestion
    assert not file_parser.check_date_input("2020-01-01")
```

File: tests/lambda_endpoints/test_path_helper.py
Comment (greglucas) [inline tests/lambda_endpoints/test_path_helper.py]: comparisons should be `assert something_is_true` or `assert not something_is_false` and not comparing against the specific object.
```suggestion
    assert FilenameParser(filename).validate_filename()
```

File: tests/lambda_endpoints/test_path_helper.py
Comment (greglucas) [inline tests/lambda_endpoints/test_path_helper.py:10]: Why do you need to input a date to the fileparser? Shouldn't this be acting on `self.date`?

File: sds_data_manager/lambda_code/SDSCode/path_helper.py
Comment (tech3371) [reply sds_data_manager/lambda_code/SDSCode/path_helper.py]: I did remove it originally based on feedback from you and Greg. But when I input correct date such as `2023105`, `strptime` accepted because it took that as valid date even though its document says otherwise.
https://docs.python.org/3/library/datetime.html#strftime-strptime-behavior

File: tests/lambda_endpoints/test_path_helper.py
Comment (greglucas) [inline tests/lambda_endpoints/test_path_helper.py]: These tests might be easier to parameterize over your assertions.
https://docs.pytest.org/en/7.1.x/example/parametrize.html

```suggestion
@pytest.mark.parametrize("filename,expected", [("imap_glows_l0_raw_20231010_20231011_v01-01.pkts", True), ...]
def test_filename_validator(filename, expected):
    assert FilenameParser(filename).validate_filename() == expected
```

**Review by greglucas — COMMENTED** [review]
I'm fine with this and won't block, but I do find this harder to follow by creating a class with a bunch of methods than individual functions would be. Since others like it this way that is fine though.

```python
def create_upload_key(fname, date_validator=is_valid_date):
    # Validate the pieces of our filename
    mission, instrument, level = fname.split()
    # call out to other validation functions
    if not date_validator(start_date):
        return False
    ...
    # assemble parts after validations

def is_valid_date(datestr, date_format="YYYYMMDD"):
    try:
        datetime.strptime(datestr, date_format)
    except ValueError:
        return False
    return True
```

File: sds_data_manager/lambda_code/SDSCode/path_helper.py
Comment (tech3371) [reply sds_data_manager/lambda_code/SDSCode/path_helper.py]: I will remove it since you guys are sure of this. As I mentioned above, `strptime` takes in seven digit date as long as it's a valid date such as `2023101`. I feel like I am missing something here.

File: tests/lambda_endpoints/test_path_helper.py
Comment (tech3371) [reply tests/lambda_endpoints/test_path_helper.py:10]: Yes. but I wrote that function to take in an input date since we need to check for both startdate and enddate. And since I wanted to check the date format only, I figure I could call the function directly instead of calling class again and again.

File: sds_data_manager/lambda_code/SDSCode/path_helper.py
Comment (maxinelasp) [reply sds_data_manager/lambda_code/SDSCode/path_helper.py]: Odd! Well, good job verifying 😆

File: sds_data_manager/lambda_code/SDSCode/path_helper.py
Comment (tech3371) [reply sds_data_manager/lambda_code/SDSCode/path_helper.py]: let me know if you like me to remove it still.

Comment (greglucas) [reply sds_data_manager/lambda_code/SDSCode/path_helper.py]: Nope, it looks like I was wrong here! Good catch on your part and sorry for both of us mentioning you could remove it 🐑 

Another option to avoid regex would be to just check the length: `if len(input_date) != 8` since the following sequence will check the numbers.

Comment (tech3371) [reply sds_data_manager/lambda_code/SDSCode/path_helper.py]: true. that's much simpler.

**Pull request merged**
